pkg/kube/krt/nestedjoin.go

```diff
@@ -0,0 +1,543 @@
+			events = append(events, Event[T]{Old: m, New: res, Event: controllers.EventUpdate})
+		}
+		handler(events)
+	case collectionMembershipEventUpdate:
+		// Get all of the elements in the old collection
+		oldItems := e.oldCollectionValue.List()
+		// Convert it to a sparse map for easy lookup
+		oldItemsMap := make(map[Key[T]]T, len(oldItems))
+		for _, i := range oldItems {
+			key := getTypedKey(i)
+			oldItemsMap[key] = i
+		}
+		// Now loop through the new collection and compare it to the old one
+		seen := sets.NewWithLength[string](len(oldItems))
+		finalEvents := make([]Event[T], 0, len(oldItems))
+		for _, i := range e.collectionValue.List() {
+			key := getTypedKey(i)
+			// If we see it in the old collection, then it's an update
+			if oldItem, ok := oldItemsMap[key]; ok {
+				seen.Insert(string(key))
+				merged := j.GetKey(string(key))
+				if merged == nil {
+					// This shouldn't happen, log it
+					log.Warnf("NestedJoinCollection: Merged item %v is nil after a collection update. Falling back to collection specific version", key)
+					merged = &i
+				}
+				// Send an update event for the merged version of this key
+				finalEvents = append(finalEvents, Event[T]{Old: &oldItem, New: merged, Event: controllers.EventUpdate})
+				// Delete it from the old items map
+				delete(oldItemsMap, key)
+			} else {
+				if seen.Contains(string(key)) {
+					// This is a duplicate item in the new collection, skip it
+					log.Warnf("NestedJoinCollection: Duplicate item %v in updated collection, skipping", key)
+					continue
+				}
+				// This is a new item
+				finalEvents = append(finalEvents, Event[T]{New: &i, Event: controllers.EventAdd})
+			}
+		}
+		// Now loop through the old items map and send delete events for any items that
+		// are no longer in the new collection
+		for _, i := range maps.SeqStable(oldItemsMap) {
+			finalEvents = append(finalEvents, Event[T]{Old: &i, Event: controllers.EventDelete})
+		}
+
+		handler(finalEvents)
+	}
+}
+
+func (j *nestedjoin[T]) RegisterBatch(f func(o []Event[T]), runExistingState bool) HandlerRegistration {
+	if j.merge == nil {
+		return j.registerBatchUnmerged(f, runExistingState)
+	}
+	syncers := make(map[collectionUID]Syncer)
+	removes := map[collectionUID]func(){}
+
+	for _, c := range j.collections.List() {
+		ic := c.(internalCollection[T])
+		reg := c.RegisterBatch(j.handleInnerCollectionEvent(f), runExistingState)
+		removes[ic.uid()] = reg.UnregisterHandler
+		syncers[ic.uid()] = reg
+	}
+	djhr := &dynamicJoinHandlerRegistration{
+		syncers: syncers,
+		removes: removes,
+	}
+
+	// We register to get notified if a collection within our set of collections is modified
+	j.registerCollectionChangeHandler(func(e collectionChangeEvent[T]) {
+		j.handleCollectionChangeEventLocked(djhr, e, f)
+	})
+
+	return djhr
+}
+
+func (j *nestedjoin[T]) handleInnerCollectionEvent(handler func(o []Event[T])) func(o []Event[T]) {
+	return func(events []Event[T]) {
```

Comment:
does the returned wrapped-handler need to hold read-lock `djhr`?

Reply:
I wouldn't think so because it the handlers aren't exposed to and don't modify the registration. Maybe I'm missing what you're saying

Reply:
yeah I got myself mixed up here; sorry.